Add sapphire docs (#171)
oasisprotocol/docs: docs/general/developer-resources/sapphire-paratime/writing-dapps-on-sapphire.md

---
description: Considerations for creating DApps on Sapphire
---

> # Creating DApps for Sapphire

Reviewer: Hmm... Emerald tutorial uses "dApps". Can we be consistent?
What would be the correct naming, dApp or Dapp :slightly_smiling_face: ?

/cc @peterjgilbert

Reviewer 2: I agree we should be consistent. Let's just use "dApp" since we're already using that everywhere else.

Author: "dapp" since that's what eth uses now. it _was_ DApp originally, though.

Author: also, if you grep the `oasis-sdk` repo, you'll find:

```
oasis-sdk on  HEAD
❯ rg "dapp"
client-sdk/ts-web/ext-utils/README.md
11:be used with DApps.
12:Prefer to have more responsibilities in the DApp, such as connecting to a
18:Prefer to have a DApp request to interact with an extension.

client-sdk/ts-web/ext-utils/docs/getting-started.md
3:## Building a DApp
5:**The sample DApp.**
6:Take a look at the sample code for the DApp side of the overall interface.
17:**Your own DApp.**
29:Take a look at the sample code for the extension side of the overall DApp
38:callbacks, which the `oasisExt` library will call when the DApp makes
41:DApp will embed, but it's fine to relay messages to a background page or
52:To try it out with the sample DApp page, do the following:
```

so it seems like `dApp` was out of place to begin with

This page mainly describes the differences between Sapphire and Ethereum
since there are a number of excellent tutorials on developing for Ethereum.

> If you don't know where to begin, the [Hardhat tutorial] and [Solidity docs]

Reviewer: Can you point to the "Writing dApps on Emerald" guide here and include links on external Hardhat tutorial / Solidity docs in the "Writing dApps on Emerald" guide, probably somewhere in the first paragraph?

Author: Sure, but I also don't think we should be maintaining a guide on writing solidity dapps. If people can't figure that out, they shouldn't be developing on Sapphire.

[Hardhat tutorial]: https://hardhat.org/tutorial
[Solidity docs]: https://docs.soliditylang.org/en/v0.8.15/solidity-by-example.html

## Oasis Consensus Layer and Sapphire ParaTime

The Oasis Network consists of the consensus layer and a number of ParaTimes.
ParaTimes are independent replicated state machines that settle transactions
using the consensus layer (to learn more, check the [Oasis Network Overview]
[overview chapter]). Sapphire is a ParaTime which implements the Ethereum
Virtual Machine (EVM).

The minimum and also expected block time in Sapphire is **6 seconds**. Any
Sapphire transaction will require at least this amount of time to be executed,
and probably no more.

ParaTimes, Sapphire included, are not allowed to directly access your tokens stored
in consensus layer accounts. You will need to _deposit_ tokens from your consensus
account to Sapphire. Consult the [How to transfer ROSE into an EVM ParaTime]
[how-to-deposit-rose] chapter to learn more.

[overview chapter]: ../../oasis-network/overview.md
[how-to-deposit-rose]: ../../manage-tokens/how-to-transfer-rose-into-evm-paratime.mdx
[Testnet faucet]: https://faucet.testnet.oasis.dev/

## Testnet and Mainnet

> The Sapphire ParaTime is currently deployed on Testnet, with Mainnet deployment planned for Q3 2022.

Reviewer 2: ```suggestion
The Sapphire ParaTime is currently deployed on Testnet, with Mainnet deployment planned for later in 2022.
```

  Unsigned calls have their sender set to the zero address. This allows contract authors
  to write getters that release secrets to authenticated callers, but without without
  requiring a transaction to be posted on-chain.

In addition to confidentiality, you get a few extra benefits including the ability to generate private
entropy, and make signatures on-chain. An example of a DApp that uses both is a HSM contract
that generates an Ethereum wallet and signs transactions sent to it via transactions.

Otherwise Sapphire is like Emerald, which is like a fast, cheap Ethereum.

## Integrating Sapphire

Once ROSE tokens are [deposited into Sapphire], it should be painless for users to begin
using DApps. To achieve this ideal user experience, we have to modify the DApp a little,
but it's made simple by our compatibility library, [`@oasislabs/sapphire-paratime`]
(coming soon). Please refer to the library docs for more information on its use.

There are compatibility layers in other languages, which may be found in [the repo].

[deposited into Sapphire]: ../../manage-tokens/how-to-transfer-rose-into-evm-paratime.mdx

> [`@oasislabs/sapphire-paratime`]: https://www.npmjs.com/@oasislabs/sapphire-paratime

Reviewer: Also, it would be better to host this under `@oasisprotocol` namespace, e.g. https://www.npmjs.com/package/@oasisprotocol/sapphire-paratime.

Reviewer 2: This package actually doesn't exist yet. I'd suggest we just remove the link until it exists.

Reviewer: Agreed

app may need to pad state items to a constant length, or use other obfuscation.
Observers may also be able to infer computation based on storage access patterns,
so you might need to obfuscate that, too.

Contract state may be made available to third parties through logs/events, or
explicit getters.

### Contract Logs

Contract logs/events (e.g., those emitted by the Solidity `emit` keyword)
are exactly like Ethereum. Data contained in events is *not* encrypted.
Precompiled contracts are available to help you encrypt data that you can
then pack into an event, however.

:::danger Unmodified contracts may leak state through logs

Base contracts like those provided by OpenZeppelin often emit logs containing
private information. If you don't know they're doing that, you might undermine
the confidentiality of your state. As a concrete example, the ERC-20 spec
requires implementers to emit an `event Transfer(from, to, amount)`, which is
obviously problematic if you're writing a confidential token. What you can
do instead is fork that contract and remove the offending emissions.

:::

### Precompiles

Sapphire provides a few extra precompiled contracts that you may find useful.

> Some of these may not be available for use initially, but all are coming soon.

Reviewer: Can you wrap this in an admonion, e.g.:

```
:::note

:::
```

so it will be more noticable?

Author: ```suggestion
:::note

Some of these may not be available for use initially, but all are coming soon.

:::
```

* `0x010000000000000000000000000000000000000000`  
  `randomBytes(uint256 numWords) returns (bytes)`  
  Generates cryptographically secure bytes. Reverts if there is not enough entropy.
* `0x010000000000000000000000000000000000000001`  
  `k256Sign(bytes32 key, bytes32 dataHash) returns (bytes32 r, bytes32 s, bytes32 v)`  
  Signs the provided data hash using the provided secp256k1 private key.
* `0x010000000000000000000000000000000000000002`  
  `deoxysii256Encrypt(bytes32 key, bytes data) returns (bytes)`  

>   Encrypts the provided data using DeoxysII with provided (random uniform) key

Reviewer: ```suggestion
  Encrypts the provided data using Deoxys-II with provided (random uniform) key
```

[Hardhat tutorial]: https://hardhat.org/tutorial
[Solidity docs]: https://docs.soliditylang.org/en/v0.8.15/solidity-by-example.html

## Oasis Consensus Layer and Sapphire ParaTime

The Oasis Network consists of the consensus layer and a number of ParaTimes.
ParaTimes are independent replicated state machines that settle transactions
using the consensus layer (to learn more, check the [Oasis Network Overview]
[overview chapter]). Sapphire is a ParaTime which implements the Ethereum
Virtual Machine (EVM).

The minimum and also expected block time in Sapphire is **6 seconds**. Any
Sapphire transaction will require at least this amount of time to be executed,
and probably no more.

ParaTimes, Sapphire included, are not allowed to directly access your tokens stored
in consensus layer accounts. You will need to _deposit_ tokens from your consensus
account to Sapphire. Consult the [How to transfer ROSE into an EVM ParaTime]
[how-to-deposit-rose] chapter to learn more.

[overview chapter]: ../../oasis-network/overview.md
[how-to-deposit-rose]: ../../manage-tokens/how-to-transfer-rose-into-evm-paratime.mdx
[Testnet faucet]: https://faucet.testnet.oasis.dev/

## Testnet and Mainnet

> The Sapphire ParaTime has two main deployments: [Sapphire Mainnet] and [Sapphire Testnet].

Reviewer 2: ```suggestion
The Sapphire ParaTime is currently deployed on Testnet, with Mainnet deployment planned for Q3 2022.
```

Like other chains' mainnet/testnet deployments, the state on each deployment is independent.
The Mainnet deployment is stable and is intended to persist state in perpetuity. The Testnet,
on the other hand, should be considered unstable software and may have its state wiped at any
time. As the name implies, only use the Testnet for testing unless you're testing how angry

> your users get when state is wiped.

Reviewer 2: ```suggestion
The Testnet should be considered unstable software and may have its state wiped at any
time. As the name implies, only use the Testnet for testing unless you're testing how angry
your users get when state is wiped.
```

:::danger Never deploy production services on Testnet

Because Testnet state can be wiped in the future, you should **never** deploy a
production service on the Testnet! Just don't do it!

:::

:::tip

For testing purposes, visit our [Testnet faucet] to obtain some TEST which you
can then use on the Sapphire Testnet to pay for gas fees. The faucet supports
sending TEST both to your consensus layer address or to your address inside the
ParaTime.

:::

[network-parameters]: ../../oasis-network/network-parameters.md
[Testnet]: ../../foundation/testnet/README.md
[Sapphire Mainnet]: ./README.mdx#mainnet

> [Sapphire Testnet]: ./README.mdx#testnet

Reviewer 2: ```suggestion
```

## Sapphire vs Ethereum

The Sapphire ParaTime is generally compatible with Ethereum, the EVM, and all of the
user and developer tooling that you already use. There are a few breaking changes,
but we think that you'll like them:

* Contract state is only visible to the contract that wrote it. With respect
  to the contract API, it's as if all state variables are declared as `private`, but
  with the further restriction that not even full nodes can read the values. Public or
  access-controlled values are provided instead through explicit getters.
* Transactions and calls are end-to-end encrypted into the contract. Only the caller
  and the contract can see the data sent to/received from the ParaTime. This ends up
  defeating most of the utility of block explorers, however.
* The `from` address using of calls is derived from a signature attached to the call.
  Unsigned calls have their sender set to the zero address. This allows contract authors
  to write getters that release secrets to authenticated callers, but without without
  requiring a transaction to be posted on-chain.

> With confidentiality, you get a few extra benefits including the ability to generate private

Reviewer 2: ```suggestion
In addition to confidentiality, you get a few extra benefits including the ability to generate private
```

entropy, and make signatures on-chain. An example of a DApp that uses both is a HSM contract
that generates an Ethereum wallet and signs transactions sent to it via transactions.

Otherwise Sapphire is like Emerald, which is like a fast, cheap Ethereum.

## Integrating Sapphire

Once ROSE tokens are [deposited into Sapphire], it should be painless for users to begin
using DApps. To achieve this ideal user experience, we have to modify the DApp a little,
but it's made simple by our compatibility library, [`@oasislabs/sapphire-paratime`].
Please refer to the library docs for more information on its use.

There are compatibility layers in other languages, which may be found in [the repo].

[deposited into Sapphire]: ../../manage-tokens/how-to-transfer-rose-into-evm-paratime.mdx

> [`@oasislabs/sapphire-paratime`]: https://www.npmjs.com/@oasislabs/sapphire-paratime

Reviewer 2: We should probably mark this as "coming soon" for now, so we can go ahead and publish these docs.

[the repo]: https://github.com/oasisprotocol/sapphire-paratime/tree/main/clients

## Writing Secure DApps

### Wallets

Sapphire is compatible with popular self-custodial wallets including MetaMask,

> Ledger, Gnosis Safe, Brave, and so forth. You can also use libraries like Web3.js

Reviewer 2: Not sure we want to mention Gnosis Safe here, since someone would need to deploy both the contract(s) and frontend in order for users to use it.

You can find the details of the Oasis Sapphire Web3 gateway
[here](/general/developer-resources/sapphire-paratime#web3-gateway).

### Transactions & Calls

Transactions and calls must be encrypted and signed for maximum security. You can
use the [`@oasislabs/sapphire-paratime`] JS package to make your life easy. It'll
handle cryptography and signing for you.

One thing to keep in mind is that `msg.sender` will be zeroed for calls that are
not signed. If you want to use `msg.sender` for access control, the call must be
signed. If you want to avoid signature popups in the user's wallet, just set the
`from` address to all zeros. The JS library will do this for you.

### Contract State

The Sapphire state model is like Ethereum's except for all state being encrypted
and not accessible to anyone except the contract. The contract, executing in an
active (attested) Oasis compute node is the only entity that can request its its
state encryption key from the Oasis key manager. Both the keys and values of the
items stored in state are encrypted, but the size of either is *not* hidden. You
app may need to pad state items to a constant length, or use other obfuscation.

> Oasis storage nodes may also infer computation based on storage access patterns,
> so you may need to obfuscate that, too.

Reviewer 2: ```suggestion
Observers may also be able to infer computation based on storage access patterns,
so you might need to obfuscate that, too.
```

Contract state may be made available to third parties through logs/events, or
explicit getters.

### Contract Logs

Contract logs/events (e.g., those emitted by the Solidity `emit` keyword)
are exactly like Ethereum. Data contained in events is *not* encrypted.
Precompiled contracts are available to help you encrypt data that you can
then pack into an event, however.

:::danger Unmodified contracts may leak state through logs

Base contracts like those provided by OpenZeppelin often emit logs containing
private information. If you don't know they're doing that, you might undermine
the confidentiality of your state. As a concrete example, the ERC-20 spec
requires implementers to emit an `event Transfer(from, to, amount)`, which is
obviously problematic if you're writing a confidential token. What you can
do instead is fork that contract and remove the offending emissions.

:::

### Precompiles

> Sapphire provides a few extra precompiled contracts that you may find useful.

Reviewer 2: These precompiles aren't available yet, right?

Author: not yet. I'll make a note of that
